clients/drcachesim/tracer/raw2trace.h

```diff
@@ -71,6 +71,128 @@ struct module_t {
     bool is_external; // If true, the data is embedded in drmodtrack custom fields.
 };
 
+/**
+ * instr_summary is a compact encapsulation of the information needed by trace conversion
+ * from decoded instructions.
+ */
+struct instr_summary_t final {
+    instr_summary_t()
+    {
+    }
+
+    /**
+     * Populates a pre-allocated instr_summary_t desc, from the instruction found at pc.
+     * Updates pc to the next instruction. Optionally logs translation details (using
+     * orig_pc and verbosity)
+     */
+    static bool
+    construct(void *dcontext, INOUT app_pc *pc, app_pc orig_pc, OUT instr_summary_t *desc,
+              uint verbosity = 0);
+
+    /**
+     * Get the pc after the instruction that was used to produce this instr_summary_t.
+     */
+    app_pc
+    next_pc() const
+    {
+        return next_pc_;
+    }
+
+private:
+    friend class raw2trace_t;
+
+    byte
+    length() const
+    {
+        return length_;
+    }
+    uint16_t
+    type() const
+    {
+        return type_;
+    }
+    uint16_t
+    prefetch_type() const
+    {
+        return prefetch_type_;
+    }
+
+    bool
+    reads_memory() const
+    {
+        return TESTANY(kReadsMemMask, packed_);
+    }
+    bool
+    writes_memory() const
+    {
+        return TESTANY(kWritesMemMask, packed_);
+    }
+    bool
+    is_prefetch() const
+    {
+        return TESTANY(kIsPrefetchMask, packed_);
+    }
+    bool
+    is_flush() const
+    {
+        return TESTANY(kIsFlushMask, packed_);
+    }
+    bool
+    is_cti() const
+    {
+        return TESTANY(kIsCtiMask, packed_);
+    }
+
+    const opnd_t &
+    src_at(size_t pos) const
+    {
+        return srcs_and_dests_[pos];
+    }
+    const opnd_t &
+    dest_at(size_t pos) const
+    {
+        return srcs_and_dests_[srcs_ + pos];
```

Comment:
Combining srcs and dsts just seems to add complexity: is there something positive coming out of it?  Maybe best to separate.

Reply:
one less large object (another std::vector is larger than the 2 bytes). This helps in an implementation where we don't allocated individual instr_summary_t on the heap, but rather bulk-allocate entire pages of them.